### ISSUE TITLE
CAPV - Too many sessions leads to 503 Error "unable to create tags manager: POST https://<SERVER>/rest/com/vmware/cis/session: 503"

### ISSUE THREAD

**Issue by PatrickLaabs** [kubernetes-sigs/cluster-api-provider-vsphere#1818]
/kind bug

**What steps did you take and what happened:**
I ran the 'Getting Started'-Guide for the vSphere CAPI Provider and was able to create a management cluster and two attached workload-Clusters.
BUT: Most of the times when creating a new cluster, or even deleting an existing cluster, 
I get the follow error when I run `clusterctl describe cluster <CLUSTERNAME>`:

```
unable to create tags manager: POST https://<SERVER>/rest/com/vmware/cis/session: 503
```

A short look inside the logs tells me:

```
2023-03-07T11:34:53.328Z | jetty-default-245         | Invoking operation get from service com.vmware.cis.session with id 3d18a644-129c-40f5-9ee1-72f35d7805ea
2023-03-07T11:34:53.329Z | jetty-default-245         | 172.30.6.33 - - [07/Mar/2023:11:34:53 +0000] "POST /rest/com/vmware/cis/session?~action=get HTTP/1.1" 401 442 "-" "infrastructure.cluster.x-k8s.io/v1beta1" 1
2023-03-07T11:34:53.464Z | jetty-default-985         | Invoking operation get from service com.vmware.cis.session with id 82c6241d-0a44-460e-be7a-d792da750a08
2023-03-07T11:34:53.464Z | jetty-default-985         | 172.30.6.33 - - [07/Mar/2023:11:34:53 +0000] "POST /rest/com/vmware/cis/session?~action=get HTTP/1.1" 401 442 "-" "infrastructure.cluster.x-k8s.io/v1beta1" 0
2023-03-07T11:34:53.681Z | jetty-default-1103        | Invoking operation get from service com.vmware.cis.session with id 0dd47265-f5ff-4dde-81d3-05342856c619
2023-03-07T11:34:53.681Z | jetty-default-1103        | 172.30.6.33 - - [07/Mar/2023:11:34:53 +0000] "POST /rest/com/vmware/cis/session?~action=get HTTP/1.1" 401 442 "-" "infrastructure.cluster.x-k8s.io/v1beta1" 0
2023-03-07T11:34:55.832Z | jetty-default-985         | Invoking operation get from service com.vmware.cis.session with id 3a3e47d1-febd-44d4-9e2f-fb953d7fd341
2023-03-07T11:34:55.833Z | jetty-default-985         | 172.30.6.33 - - [07/Mar/2023:11:34:55 +0000] "POST /rest/com/vmware/cis/session?~action=get HTTP/1.1" 401 442 "-" "infrastructure.cluster.x-k8s.io/v1beta1" 2
2023-03-07T11:35:00.170Z | jetty-default-245         | Invoking operation get from service com.vmware.cis.session with id 9666fa1b-c87a-40dc-836e-0ab63905406c
2023-03-07T11:35:00.171Z | jetty-default-245         | 172.30.6.33 - - [07/Mar/2023:11:35:00 +0000] "POST /rest/com/vmware/cis/session?~action=get HTTP/1.1" 200 128 "-" "infrastructure.cluster.x-k8s.io/v1beta1" 2
2023-03-07T11:35:00.350Z | jetty-default-246         | Invoking operation get from service com.vmware.cis.session with id ff7acc7d-f646-46cd-9648-def2dcefb301
2023-03-07T11:35:00.351Z | jetty-default-246         | 172.30.6.33 - - [07/Mar/2023:11:35:00 +0000] "POST /rest/com/vmware/cis/session?~action=get HTTP/1.1" 200 128 "-" "infrastructure.cluster.x-k8s.io/v1beta1" 1
2023-03-07T11:35:03.743Z | jetty-default-245         | Invoking operation get from service com.vmware.cis.session with id b01effe1-a3da-4496-9aef-7710b41ed428
2023-03-07T11:35:03.743Z | jetty-default-245         | 172.30.6.33 - - [07/Mar/2023:11:35:03 +0000] "POST /rest/com/vmware/cis/session?~action=get HTTP/1.1" 401 442 "-" "infrastructure.cluster.x-k8s.io/v1beta1" 1
2023-03-07T11:35:03.874Z | jetty-default-246         | Invoking operation get from service com.vmware.cis.session with id 17ac6e42-5119-44a1-9caa-90eda95740b6
2023-03-07T11:35:03.874Z | jetty-default-246         | 172.30.6.33 - - [07/Mar/2023:11:35:03 +0000] "POST /rest/com/vmware/cis/session?~action=get HTTP/1.1" 401 442 "-" "infrastructure.cluster.x-k8s.io/v1beta1" 0
2023-03-07T11:35:04.068Z | jetty-default-1103        | Invoking operation get from service com.vmware.cis.session with id c4392f64-fad0-41d7-8453-3229246f8555
2023-03-07T11:35:04.068Z | jetty-default-1103        | 172.30.6.33 - - [07/Mar/2023:11:35:04 +0000] "POST /rest/com/vmware/cis/session?~action=get HTTP/1.1" 401 442 "-" "infrastructure.cluster.x-k8s.io/v1beta1" 0
2023-03-07T11:35:06.715Z | jetty-default-985         | Invoking operation get from service com.vmware.cis.session with id f967dd2f-50f8-42da-8307-7cc8039bd30c
2023-03-07T11:35:06.715Z | jetty-default-985         | 172.30.6.33 - - [07/Mar/2023:11:35:06 +0000] "POST /rest/com/vmware/cis/session?~action=get HTTP/1.1" 200 128 "-" "infrastructure.cluster.x-k8s.io/v1beta1" 1
2023-03-07T11:35:06.884Z | jetty-default-245         | Invoking operation get from service com.vmware.cis.session with id 848f1fb9-3056-4cc8-a1ac-da1367ee4a76
```

And at the end my vSphere blocks the user from accessing the API:
```
2023-03-07T11:40:37.663Z | WARN  | sso6                      | BaseSessionImpl                | User sessions count is limited to 550. Existing sessions are 550 for user <USER>@<DOMAIN>. Please retry the login operation
```

We increased the limit from 500 to 2000, but the problem still persists. 
It is not thrown directly anymore, just takes a little longer, but provisioning a new workload cluster with 1x controlplane and 3x worker nodes fails.
So I either have to wait, that my limit is being reseted or keep increasing the ratelimit.. but I would guess, that this might not be the right solution to deal with this kind of error.

**What did you expect to happen:**
I am not expecting so many API calls. So many, that my vCenter blocks my Useraccount for the API.
I do expected, that this might need a few calls against the api, but not above 2000.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]
I was going through some changes, and found out, that the keep-alive option should be turned on by default.

**Environment:**

- Cluster-api-provider-vsphere version:   v1.5.3
- Kubernetes version: (use `kubectl version`):  v1.25.6 (OVA Image from the documentation of this provider)
- OS (e.g. from `/etc/os-release`):  `VERSION_ID="22.04"
VERSION="22.04.1 LTS (Jammy Jellyfish)"`

**Comment by PatrickLaabs** [comment]
I perfomed an upgrade on my management cluster to the latest version of the vsphere-provider.
Version 1.6.0 is now installed.

I will update this issue as soon as I get new informations on this. (My User is currently locked) 😞

**Comment by srm09** [comment]
I am starting to think we should re-introduce the keep alive flag, and also have an idle timeout set on the sessions. Irrespective of the keep alive flag, the idle timeout would clean up sessions that have been inactive for a particular period of time. That way, we make sure that the existing sessions (say, with the keep alive flag enabled) do not stay on the cluster indefinitely (case wherein a particular cluster for which the session was created ends up getting deleted in which that session will be unused and should be cleaned up)

**Comment by PatrickLaabs** [comment]
After upgrading the vSphere-Provider components on the management cluster, and generating a new manifest for a workload cluster and deploying the new manifest the error still remains. 

I had to stop the capv-controller today, since our vCenter crashed twice 😄 

Tomorrow I will delete everything, generate new manifests with the new Codebase and give it another shot. 

I will update this issue as soon as I got new insights on this.

**Comment by PatrickLaabs** [comment]
Update:

Running the latest CAPV-Component is defintily worth it. The Session-Leak seems to be fixed.

But:
We also discovered, that we were missing a role for our user, that we are using with capi.
We allowed him to view active sessions, and disconnect them.

So, i guess, it should've worked also with the v1.5.3 Release.

Maybe a documentation for setting up a dedicated user inside vsphere, with all the needed roles/access settings might be interesting, or?

Best
Patrick
